Fix Multizone pv creation on GCE

gnufied opened the pull request:
When Multizone is enabled static PV creation on GCE
fails because Cloud provider configuration is not
available in admission plugins.

cc @derekwaynecarr @childsb

gnufied committed:
Fix Multizone pv creation on GCE

When Multizone is enabled static PV creation on GCE
fails because Cloud provider configuration is not
available in admission plugins.

gnufied commented on plugin/pkg/admission/persistentvolume/label/admission_test.go:
```diff
@@ -75,6 +77,14 @@ func mockVolumeLabels(labels map[string]string) *mockVolumes {
+        `
+	return strings.NewReader(gceConfigData)
+}
+
```
Ignore this for now. I added this in preparation for adding some tests in admission plugin as well.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/40537)
<!-- Reviewable:end -->

ncdc commented on plugin/pkg/admission/persistentvolume/label/admission.go:
```diff
@@ -57,6 +58,10 @@ func NewPersistentVolumeLabel() *persistentVolumeLabel {
 	}
 }
 
+func (l *persistentVolumeLabel) SetCloudConfig(cloudConfig io.Reader) {
```
Remind me if this is required, or can we run this plugin without it?

gnufied replied:
It does run without it. For AWS  relevant cloud configuration is either picked from environment variable or in case of GCE - instance must have API access.

ncdc replied:
Ok thanks. If it were required, I was going to suggest you implement `Validate`, but it doesn't sound like we need that.

gnufied commented:
cc @kubernetes/sig-storage-pr-reviews

gnufied commented:
@k8s-bot kops aws e2e test this

derekwaynecarr commented on federation/cmd/federation-apiserver/app/server.go:
```diff
@@ -159,7 +159,7 @@ func Run(s *options.ServerRunOptions) error {
 
 	admissionControlPluginNames := strings.Split(s.GenericServerRunOptions.AdmissionControl, ",")
-	pluginInitializer := kubeapiserveradmission.NewPluginInitializer(client, sharedInformers, apiAuthorizer)
+	pluginInitializer := kubeapiserveradmission.NewPluginInitializer(client, sharedInformers, apiAuthorizer, nil)
```
federation has the cloud provider options.  i would include support for it here too.

gnufied replied:
fixed.

derekwaynecarr reviewed (changes requested):
please add the same support for kubefed

gnufied commented:
Rebased and fixed comment about federation api server. Also manually verified on GCE that this change indeed enables creation of PVs in zones (not regions to be clear) other than master.

@derekwaynecarr @childsb ptal.

gnufied commented:
@k8s-bot kops aws e2e test this

jsafrane commented:
@k8s-bot kops aws e2e test this

gnufied commented:
Flaked on https://github.com/kubernetes/kubernetes/issues/37891

derekwaynecarr commented:
/approve

derekwaynecarr commented:
@k8s-bot test this

gnufied commented:
@k8s-bot test this

gnufied commented:
@k8s-bot test this

ncdc commented:
@k8s-bot kops aws e2e test this

gnufied commented:
@k8s-bot test this

gnufied commented:
@k8s-bot kops aws e2e test this

smarterclayton commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -299,7 +299,18 @@ func Run(s *options.ServerRunOptions) error {
+	if s.CloudProvider.CloudConfigFile != "" {
+		cloudConfigReader, err = os.Open(s.CloudProvider.CloudConfigFile)
+		if err != nil {
+			glog.Errorf("Error opening cloud configuration file %s: %#v", s.CloudProvider.CloudConfigFile, err)
```
Why is this not a fatal error?

gnufied replied:
Because I intended this change to be backward compatible in the sense - a misconfigured apiserver shouldn't start crashing after this change.

It could be argued that, may be it should crash if cloud provider configuration is indeed broken but I thought it is best to least obstrusive.

smarterclayton replied:
If someone passes the cloud config argument and it's not a file, that should already have been an error.  It is better to treat this as a hard error to prevent future debugging and head scratching of "but i specified the cloud provider, why isn't it being used?".

gnufied replied:
Okay but api-server already seems to have a contract that, "I will not panic if I can't configure Cloudprovider properly". See - https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/server.go#L158

If we were to change that contract, I do not want to do this as part of this particular fix.  We have already seen plenty of head scratching bugs where kubernetes wasn't working properly because cloudprovider wasn't configured properly and rather than complaining loudly kubernetes was silently logging error messages. So I completely agree with the idea of logging a fatal error, but I think we should do in both places not just here.

I am bit hesitant to change logging in both places though. I don't know enough about first initialization and why we aren't logging that as fatal error.

gnufied replied:
fixed. ptal

smarterclayton commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -299,7 +299,18 @@ func Run(s *options.ServerRunOptions) error {
+		if err != nil {
+			glog.Errorf("Error opening cloud configuration file %s: %#v", s.CloudProvider.CloudConfigFile, err)
+		}
+		if cloudConfigReader != nil {
```
Passing a reader is really ugly, I'd recommend you pass the bytes down.  Just because cloud config reader accepts it today doesn't mean it would continue.  This will be a lot cleaner as ReadAll(), and passing a buffer to the deep initializer.

gnufied replied:
fixed.

saad-ali commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -299,7 +299,15 @@ func Run(s *options.ServerRunOptions) error {
+	var cloudConfig []byte
+
+	if s.CloudProvider.CloudConfigFile != "" {
+		cloudConfig, err = ioutil.ReadFile(s.CloudProvider.CloudConfigFile)
```
You might be able to use `cloudprovider.InitCloudProvider(...)` (which reads the config file) instead of reading the config file on your own here.

gnufied replied:
How would that work for federation api server? which takes same set of arguments and initialzies admission controller in similar way.

Currently federation api server seems to be simply throwing away cloud config. options etc but it does accept them same as kube-apiserver.

saad-ali replied:
Instead of opening the file and passing the contents around, pass the config file path around. Then inside your `SetCloudConfig(...)` method use the config file path and reuse `InitCloudProvider(...)`, that way you have only one place responsible for opening the config file.

msau42 replied:
I also like this option of passing the file path.

gnufied replied:
Okay passing the file path is not enough to centralize the initialization of cloudprovider in one place (namely `SetCloudConfig`), we would need name of provider too, so we would have to pass 2 params? :( The problem with passing 2 additional params to admission plugin initialize is - now it will have 5 parameters. Nothing too bad but not clean either.

The initialization of cloudprovider in admission plugin is bit adhoc - https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/persistentvolume/label/admission.go#L74 and I don't like it but I am thinking a proper fix is to simply pass initialized `cloudprovider.Interface` object directly into the plugin rather than pass ConfigPath etc. 

So I tried doing that and it has some spill over effect - like we don't always have cloudprovider initialized even in api server and I need to refactor apiserver some more to make that happen. 

At this late in 1.6 cycle, I am bit hesitant to make bigger changes and I can confirm with some certainity that fix indeed works and I was able to create PVs in multizones after this change.

@saad-ali @msau42 I think the code here was already bit non-ideal. I would like to talk to @jsafrane and understand why he didn't pass cloudprovider object directly and used ad-hoc cloudprovider initialization.

In the meanwhile, I think we should merge this - so as underlying problem is fixed. I can work on cleaning it up.

saad-ali replied:
Ack, just open up an issue as a reminder to do this clean up.

SInce this is a bug fix, post-code freeze merge is fine (marking it for 1.6).

Will leave review and signoff to @msau42

jsafrane replied:
The admission plugin is ugly hack, I tried to make it use volume plugins in https://github.com/kubernetes/kubernetes/pull/21386 and I failed at that time. With  https://github.com/kubernetes/kubernetes/pull/39425 it could be easier to use volume plugins in API server, however it's 1.7 item (and it's not on my priority list...)

Out of curiosity, why is federation API server cloud-aware? Individual federated API severs can be in different clouds, can't they?

gnufied replied:
yeah federation just takes same arguments as api-server. I think we should selectively edit out options we don't need in federation server. cc @madhusudancs 

Also, I will open a bug to clean the implementation. thanks

madhusudancs replied:
@gnufied cloud provider config doesn't make any sense in the federation API server because a federation API server can be in one of the many clusters, potentially on different providers, it can be moved around or have multiple instances in HA configuration where different instances are on different providers. It's Ok to just completely remove that flag from the federation API server for now.

msau42 commented on plugin/pkg/admission/persistentvolume/label/admission.go:
```diff
@@ -131,7 +137,11 @@ func (l *persistentVolumeLabel) getEBSVolumes() (aws.Volumes, error) {
+		if len(l.cloudConfig) > 0 {
+			cloudConfigReader = bytes.NewReader(l.cloudConfig)
+		}
+		cloudProvider, err := cloudprovider.GetCloudProvider("aws", cloudConfigReader)
```
Is it possible to process the config file outside of the plugin, and pass in the cloudProvider object instead?  I'm not sure if there's a function to get the current cloud provider?

gnufied replied:
It is possible at least for kube-apiserver, since we already have a cloud object initialized there. It wouldn't work for federation api server though. 

What are we looking to gain by passing cloudprovider object instead of configuration bytes?

saad-ali replied:
Ideally the mechanism by which to read/write the config file should be in a single place (`cloudprovider` lib makes the most sense) instead of duplicated and sprinkled throughout the code.

gnufied replied:
okay, that sounds fair. However, federation api sever is in this weird place where it accepts same set of arguments as api server but sometimes doesn't do anything with them. 

I do not want to initialize cloudprovider in federation server as part of this change because I don't know what their expectations are. If we are going to directly pass cloudprovider then I will pass `nil` from federation apiserver until they decide to initialize cloudprovider in their code and  it becomes available. Does it sound good?

saad-ali replied:
SGTM

saad-ali commented on plugin/pkg/admission/persistentvolume/label/admission.go:
```diff
@@ -57,6 +59,10 @@ func NewPersistentVolumeLabel() *persistentVolumeLabel {
 	}
 }
 
+func (l *persistentVolumeLabel) SetCloudConfig(cloudConfig []byte) {
```
Add a compile time interface compliance check:

 ```GO
var _ WantsCloudConfig = &persistentVolumeLabel{}
```

gnufied replied:
ack

msau42 commented:
/lgtm

gnufied commented:
The failures are here because of unable to pull images from registry:

```
228 09:56:28.955] +++ [0228 09:56:28] Deleting docker image gcr.io/google_containers/kube-proxy:96233fe3a9680a19d48abe3ac2fe1570
I0228 09:56:29.010] Untagged: gcr.io/google_containers/kube-proxy:96233fe3a9680a19d48abe3ac2fe1570
I0228 09:56:29.010] Deleted: 67db50318c87bd62318d488d1fc5daa21da3e203afbc38a903234becafb2975b
W0228 10:03:48.644] Could not reach any registry endpoint
W0228 10:03:48.652] !!! [0228 10:03:48] Call tree:
W0228 10:03:48.653] !!! [0228 10:03:48]  1: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:218 kube::release::create_docker_images_for_server(...)
W0228 10:03:48.655] !!! [0228 10:03:48]  2: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:94 kube::release::package_server_tarballs(...)
W0228 10:03:48.656] !!! [0228 10:03:48]  3: build/release.sh:45 kube::release::package_tarballs(...)
W0228 10:03:48.764] Could not reach any registry endpoint
W0228 10:03:48.770] !!! [0228 10:03:48] Call tree:
W0228 10:03:48.772] !!! [0228 10:03:48]  1: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:218 kube::release::create_docker_images_for_server(...)
W0228 10:03:48.773] !!! [0228 10:03:48]  2: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:94 kube::release::package_server_tarballs(...)
W0228 10:03:48.775] !!! [0228 10:03:48]  3: build/release.sh:45 kube::release::package_tarballs(...)
W0228 10:03:49.211] Could not reach any registry endpoint
W0228 10:03:49.217] !!! [0228 10:03:49] Call tree:
W0228 10:03:49.218] !!! [0228 10:03:49]  1: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:218 kube::release::create_docker_images_for_server(...)
W0228 10:03:49.220] !!! [0228 10:03:49]  2: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:94 kube::release::package_server_tarballs(...)
W0228 10:03:49.221] !!! [0228 10:03:49]  3: build/release.sh:45 kube::release::package_tarballs(...)
W0228 10:03:49.322] Could not reach any registry endpoint
W0228 10:03:49.328] !!! [0228 10:03:49] Call tree:
W0228 10:03:49.330] !!! [0228 10:03:49]  1: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:218 kube::release::create_docker_images_for_server(...)
W0228 10:03:49.331] !!! [0228 10:03:49]  2: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:94 kube::release::package_server_tarballs(...)
W0228 10:03:49.333] !!! [0228 10:03:49]  3: build/release.sh:45 kube::release::package_tarballs(...)
W0228 10:03:49.335] !!! [0228 10:03:49] previous Docker build failed
W0228 10:03:49.338] !!! [0228 10:03:49] Call tree:
W0228 10:03:49.340] !!! [0228 10:03:49]  1: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:218 kube::release::create_docker_images_for_server(...)
W0228 10:03:49.341] !!! [0228 10:03:49]  2: /var/lib/jenkins/workspace/pull-kubernetes-e2e-gce/go/src/k8s.io/kubernetes/build/lib/release.sh:94 kube::release::package_server_tarballs(...)
W0228 10:03:49.342] !!! [0228 10:03:49]  3: build/release.sh:45 kube::release::package_tarballs(...)
W0228 10:03:49.344] !!! [0228 10:03:49] previous tarball phase failed
```

gnufied commented:
@k8s-bot test this

gnufied commented:
@k8s-bot bazel test this

gnufied commented:
@k8s-bot kops aws e2e test this

msau42 commented:
/lgtm

deads2k commented:
/approve

gnufied commented:
@k8s-bot gci gke e2e test this

gnufied commented:
@k8s-bot gci gke e2e test this

gnufied commented:
@k8s-bot test this

gnufied commented:
@k8s-bot test this

saad-ali commented:
We should cherry pick fix to 1.5 branch after it merges as well.

gnufied commented:
@k8s-bot test this
